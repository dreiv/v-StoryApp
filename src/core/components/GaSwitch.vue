<script setup lang="ts">
import { CheckIcon } from 'lucide-vue-next'
import { useFormInput, type FormInputProps } from '../composables/useFormInput'

defineOptions({ inheritAttrs: false })
const props = defineProps<FormInputProps>()
const { classes, aria } = useFormInput(props)
const model = defineModel<boolean>()
</script>

<template>
  <label :class="classes" v-bind="aria">
    <input type="checkbox" :class="$style.native" v-model="model" v-bind="$attrs" />
    <div :class="$style.marker">
      <CheckIcon :class="$style.checked" :size="16" />
    </div>

    <span :class="$style.label">
      <slot>{{ label }}</slot>
    </span>
  </label>
</template>

<style module>
.input {
  display: inline-flex;
  position: relative;
  gap: var(--ga-size-12);
  user-select: none;
}

.native,
.marker {
  width: var(--ga-size-48);
  height: var(--ga-size-24);
}

.marker {
  display: flex;
  position: absolute;
  top: 0;
  left: 0;
  align-items: center;

  border: 1px solid var(--ga-color-border-action);
  border-radius: var(--ga-radius-round);

  padding: 0 var(--ga-size-4);
  pointer-events: none;
  color: var(--ga-color-icon-on-primary);

  & > .checked {
    display: none;
  }

  &::after {
    position: absolute;
    top: calc(50% - 9px);

    border-radius: var(--ga-radius-round);
    background-color: var(--ga-color-border-action);

    width: 18px;
    height: 18px;
    content: '';
  }
}

.native {
  flex-shrink: 0;
  opacity: 0;
  cursor: pointer;

  &:focus-visible + .marker {
    outline: 2px solid var(--ga-color-border-focus);
    outline-offset: 2px;
  }

  &:hover:enabled {
    &:not(:checked) + .marker {
      border-color: var(--ga-color-border-action-hover);
      background-color: var(--ga-color-surface-action-hover-2);
    }

    &:checked + .marker {
      background-color: var(--ga-color-surface-action-hover);

      &::after {
        background-color: var(--ga-color-surface-action-hover-2);
      }
    }
  }

  &:checked + .marker {
    background-color: var(--ga-color-surface-action);

    > .checked {
      display: block;
    }

    &::after {
      transform: translateX(var(--ga-size-20));
      background-color: var(--ga-color-surface-primary);
    }
  }

  &:disabled {
    cursor: not-allowed;

    + .marker {
      border-color: var(--ga-color-border-disabled);
      background-color: var(--ga-color-surface-disabled);
      color: var(--ga-color-icon-on-disabled);

      &::after {
        background-color: var(--ga-color-icon-on-disabled);
      }
    }

    ~ .label {
      color: var(--ga-color-text-disabled);
    }

    &:checked ~ .label {
      color: var(--ga-color-text-disable-selected);
    }
  }
}

.error {
  > .marker {
    border-color: var(--ga-color-border-error);
    background-color: var(--ga-color-surface-error);
    color: var(--ga-color-icon-error);

    &::after {
      background-color: var(--ga-color-icon-error);
    }
  }

  > .native:enabled {
    &:checked + .marker {
      background-color: var(--ga-color-surface-error);

      &::after {
        background-color: var(--ga-color-icon-error);
      }
    }

    &:hover + .marker {
      border-color: var(--ga-color-border-error-hover);
      background-color: var(--ga-color-surface-action-hover-2);

      &::after {
        background-color: var(--ga-color-icon-error);
      }
    }
  }
}

.label {
  min-height: var(--ga-size-24);
  font-size: var(--text-md);
  line-height: var(--text-lg--line-height);

  &:empty {
    display: none;
  }
}
</style>
