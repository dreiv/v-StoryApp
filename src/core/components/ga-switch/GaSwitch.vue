<script setup lang="ts">
import { useTemplateRef } from 'vue'
import { CheckIcon } from 'lucide-vue-next'
import { useFormInput, type FormInputProps } from '@/core/composables/useFormInput'

defineOptions({ inheritAttrs: false })
const props = defineProps<FormInputProps>()

const { classes, aria } = useFormInput(props)
const inputRef = useTemplateRef<HTMLElement>('inputRef')
const model = defineModel<boolean>()

defineExpose({ inputRef })
</script>

<template>
  <label :class="classes" v-bind="aria">
    <input ref="inputRef" type="checkbox" :class="$style.native" v-model="model" v-bind="$attrs" />
    <div :class="$style.marker">
      <CheckIcon :class="$style.checked" :size="16" />
    </div>

    <span :class="$style.label">
      <slot>{{ label }}</slot>
    </span>
  </label>
</template>

<style module>
.input {
  display: inline-flex;
  position: relative;
  gap: var(--ga-size-spacing-04);
  user-select: none;
}

.native,
.marker {
  width: var(--ga-size-spacing-09);
  height: var(--ga-size-spacing-06);
}

.marker {
  display: flex;
  position: absolute;
  top: 0;
  left: 0;
  align-items: center;

  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-action);
  border-radius: var(--ga-radius-round);

  padding-inline: var(--ga-size-spacing-02);
  pointer-events: none;
  color: var(--ga-color-icon-on-primary);

  & > .checked {
    display: none;
  }

  &::after {
    position: absolute;
    top: calc(50% - 9px); /* TODO: fix */

    border-radius: var(--ga-radius-round);
    background-color: var(--ga-color-border-action);

    width: 18px;
    height: 18px;
    content: '';
  }
}

.native {
  flex-shrink: 0;
  opacity: 0;
  cursor: pointer;

  &:focus-visible + .marker {
    outline: var(--ga-size-border-width-md) solid var(--ga-color-border-focus);
    outline-offset: var(--ga-size-spacing-01);
  }

  &:hover:enabled {
    &:not(:checked) + .marker {
      border-color: var(--ga-color-border-action-hover);
      background-color: var(--ga-color-surface-action-hover-2);
    }

    &:checked + .marker {
      background-color: var(--ga-color-surface-action-hover);

      &::after {
        background-color: var(--ga-color-surface-action-hover-2);
      }
    }
  }

  &:checked + .marker {
    background-color: var(--ga-color-surface-action);

    > .checked {
      display: block;
    }

    &::after {
      transform: translateX(20px); /* TODO: fix */
      background-color: var(--ga-color-surface-primary);
    }
  }

  &:disabled {
    cursor: not-allowed;

    + .marker {
      border-color: var(--ga-color-border-disabled);
      background-color: var(--ga-color-surface-disabled);
      color: var(--ga-color-icon-on-disabled);

      &::after {
        background-color: var(--ga-color-icon-on-disabled);
      }
    }

    ~ .label {
      color: var(--ga-color-text-disabled);
    }
    &:checked ~ .label {
      color: var(--ga-color-text-disable-selected);
    }
  }
}

.error {
  > .marker {
    border-color: var(--ga-color-border-error);
    background-color: var(--ga-color-surface-error);
    color: var(--ga-color-icon-error);

    &::after {
      background-color: var(--ga-color-icon-error);
    }
  }
  > .native:enabled {
    &:checked + .marker {
      background-color: var(--ga-color-surface-error);

      &::after {
        background-color: var(--ga-color-icon-error);
      }
    }

    &:hover + .marker {
      border-color: var(--ga-color-border-error-hover);
      background-color: var(--ga-color-surface-action-hover-2);

      &::after {
        background-color: var(--ga-color-icon-error);
      }
    }
  }
}

.label {
  font-weight: var(--ga-font-weight-normal);
  font-size: var(--ga-text-md-font-size);
  line-height: var(--ga-text-md-line-height);
  font-family: var(--ga-font-family-primary);
  letter-spacing: var(--ga-text-md-tracking);

  &:empty {
    display: none;
  }
}
</style>
