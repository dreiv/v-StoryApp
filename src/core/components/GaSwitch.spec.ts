import { shallowMount, type VueWrapper } from '@vue/test-utils'
import GaSwitch from './GaSwitch.vue'

vi.mock('../composables/useFormInput')
vi.mock('../constants')

describe('Switch', () => {
  let wrapper: VueWrapper

  beforeEach(() => {
    wrapper = shallowMount(GaSwitch)
  })

  it('should render correctly with default props', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly with label prop', () => {
    wrapper = shallowMount(GaSwitch, { props: { label: 'Toggle feature' } })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toContain('Toggle feature')
  })

  it('should render correctly with default slot', () => {
    wrapper = shallowMount(GaSwitch, { slots: { default: 'Custom Label Content' } })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toContain('Custom Label Content')
  })

  it('should render with an error prop', async () => {
    await wrapper.setProps({ error: true, errorMessage: 'Test Error' })
    const label = wrapper.find('label')

    expect(wrapper.html()).toMatchSnapshot()
    expect(label.classes('mock-error')).toBe(true)
    expect(label.attributes('aria-invalid')).toBe('true')
    expect(label.attributes('aria-errormessage')).toBe('Test Error')
  })

  it('should emit update:modelValue event when switch is clicked', async () => {
    const input = wrapper.find('input')
    await input.setValue(true)

    expect(wrapper.emitted('update:modelValue')).toBeTruthy()
    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([true])
  })

  it('should toggle the switch state when clicked multiple times', async () => {
    const input = wrapper.find('input')

    await input.setValue(true)
    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([true])

    await input.setValue(false)
    expect(wrapper.emitted('update:modelValue')?.[1]).toEqual([false])
  })

  it('should be disabled when disabled prop is true', async () => {
    await wrapper.setProps({ disabled: true })
    const input = wrapper.find('input')

    expect(input.element.disabled).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
