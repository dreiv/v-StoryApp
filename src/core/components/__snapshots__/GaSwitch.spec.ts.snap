// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Switch > should be disabled when disabled prop is true 1`] = `
"<label class="mock-input"><input type="checkbox" class="native" disabled="">
  <div class="marker">
    <anonymous-stub class="checked" size="16"></anonymous-stub>
  </div><span class="label"></span>
</label>"
`;

exports[`Switch > should render correctly in 'disabled' state 1`] = `
"<label class="mock-input"><input type="checkbox" class="native" disabled="">
  <div class="marker">
    <anonymous-stub class="checked" size="16"></anonymous-stub>
  </div><span class="label"></span>
</label>"
`;

exports[`Switch > should render correctly in 'error' state 1`] = `
"<label class="mock-input mock-error" aria-invalid="true" aria-errormessage="Test Error"><input type="checkbox" class="native">
  <div class="marker">
    <anonymous-stub class="checked" size="16"></anonymous-stub>
  </div><span class="label"></span>
</label>"
`;

exports[`Switch > should render correctly with default props 1`] = `
"<label class="mock-input"><input type="checkbox" class="native">
  <div class="marker">
    <anonymous-stub class="checked" size="16"></anonymous-stub>
  </div><span class="label"></span>
</label>"
`;

exports[`Switch > should render correctly with default slot 1`] = `
"<label class="mock-input"><input type="checkbox" class="native">
  <div class="marker">
    <anonymous-stub class="checked" size="16"></anonymous-stub>
  </div><span class="label">Custom Label Content</span>
</label>"
`;

exports[`Switch > should render correctly with label prop 1`] = `
"<label class="mock-input"><input type="checkbox" class="native">
  <div class="marker">
    <anonymous-stub class="checked" size="16"></anonymous-stub>
  </div><span class="label">Toggle feature</span>
</label>"
`;

exports[`Switch > should render with an error prop 1`] = `
"<label class="mock-input mock-error" aria-invalid="true" aria-errormessage="Test Error"><input type="checkbox" class="native">
  <div class="marker">
    <anonymous-stub class="checked" size="16"></anonymous-stub>
  </div><span class="label"></span>
</label>"
`;
